{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook 11.2: *denovo* PacBio genome assembly \n",
    "\n",
    "\n",
    "### Learning objectives: \n",
    "\n",
    "By the end of this notebook you will:\n",
    "\n",
    "+ Experience running PacBio long-read assembly.\n",
    "+ Compare long read assembly to short-read and hybrid assemblies from notebook 1.\n",
    "+ Learn about alternative assembly software.\n",
    "\n",
    "\n",
    "### Assigned readings: \n",
    "\n",
    "**Read these papers carefully, we will discuss in class:**\n",
    "\n",
    "+ Li, Fay-Wei, and Alex Harkess. “A Guide to Sequence Your Favorite Plant Genomes.” Applications in Plant Sciences 6, no. 3 (n.d.): e1030. https://doi.org/10.1002/aps3.1030.\n",
    "\n",
    "\n",
    "\n",
    "+ Sumpter, Nicholas, Margi Butler, and Russell Poulter. “Single-Phase PacBio De Novo Assembly of the Genome of the Chytrid Fungus Batrachochytrium Dendrobatidis, a Pathogen of Amphibia.” Microbiology Resource Announcements 7, no. 21 (November 2018). https://doi.org/10.1128/MRA.01348-18.\n",
    "\n",
    "\n",
    "\n",
    "**Please skim the paper below, you do not need to read it in detail:** \n",
    "+ Giordano, Francesca, Louise Aigrain, Michael A. Quail, Paul Coupland, James K. Bonfield, Robert M. Davies, German Tischler, et al. “De Novo Yeast Genome Assemblies from MinION, PacBio and MiSeq Platforms.” Scientific Reports 7, no. 1 (June 21, 2017): 3935. https://doi.org/10.1038/s41598-017-03996-z.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# conda install bioconda::miniasm\n",
    "# conda install bioconda::minimap2\n",
    "# conda install bioconda::canu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Batrachochytrium dendrobatidis genome assembly\n",
    "\n",
    "In this notebook we will continue to examine the published genome assembly data for *Bd* (https://www.ncbi.nlm.nih.gov/pubmed/30533847) that was recently assembled with PacBio reads. In the last notebook we assembled a related genome using short-read and hybrid short-read + PacBio data. Here we will assemble the genome using only PacBio long reads. In terms of the algorithms involved in assembly this will employ a quite different approach from the de Bruijn graph based approach used in the first notebook. We will try two different approaches, a very fast assembler called `miniasm`, and a much slower but more accurate method called `canu`. The latter is the method that the authors used in the study. Both are considered OLC assemblers: overlap-layout-consensus. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Miniasm assembly\n",
    "\n",
    "The miniasm assembly is performed in two steps. We first use the tool `minmap2` to calculate the overlaps among all of the pacbio reads in the fastq data file. Then we input this overlap information along with the raw data to `miniasm` which will infer a graph assembly -- the best way to uniquely connect all overlapping sequences. This involves splitting ambiguous loops in the graph and calculating the shortest path through the graph. The result can be converted to a final fasta file representing the consensus assembled contigs. \n",
    "\n",
    "\n",
    "### Overlap: get overlap of PacBio reads\n",
    "Here we align all of the PacBio reads to each other. For very large datasets this can be time consuming. New algorithms (e.g., `minimap2` have *dramatically* improved the speed of this step recently. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#minimap2 -x ava-pb -t 20 SRR7825134.fastq SRR7825134.fastq > SRR7825134.paf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This returns a result in the form of a [pairwise alignment format (.paf)](https://github.com/lh3/miniasm/blob/master/PAF.md) file. To avoid you having to compute this file I uploaded the first 20 lines of the file to a URL so you can easily look at it. Because the format is tabular we can load it as a DataFrame with pandas to look at it below. This file contains information about the positions where long reads in the data overlap each other. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>729</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.66305</td>\n",
       "      <td>9555</td>\n",
       "      <td>1321</td>\n",
       "      <td>8140</td>\n",
       "      <td>1083</td>\n",
       "      <td>6903</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:74</td>\n",
       "      <td>s1:i:1008</td>\n",
       "      <td>dv:f:0.1064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>85</td>\n",
       "      <td>5480</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.135869</td>\n",
       "      <td>13250</td>\n",
       "      <td>7604</td>\n",
       "      <td>12972</td>\n",
       "      <td>995</td>\n",
       "      <td>5486</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:69</td>\n",
       "      <td>s1:i:935</td>\n",
       "      <td>dv:f:0.1004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>729</td>\n",
       "      <td>7404</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.50128</td>\n",
       "      <td>16210</td>\n",
       "      <td>653</td>\n",
       "      <td>7265</td>\n",
       "      <td>970</td>\n",
       "      <td>6743</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:54</td>\n",
       "      <td>s1:i:913</td>\n",
       "      <td>dv:f:0.1179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>97</td>\n",
       "      <td>7455</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR7825134.7296</td>\n",
       "      <td>13396</td>\n",
       "      <td>5948</td>\n",
       "      <td>13367</td>\n",
       "      <td>950</td>\n",
       "      <td>7458</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:59</td>\n",
       "      <td>s1:i:910</td>\n",
       "      <td>dv:f:0.1181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>55</td>\n",
       "      <td>6643</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR7825134.119454</td>\n",
       "      <td>10268</td>\n",
       "      <td>218</td>\n",
       "      <td>6869</td>\n",
       "      <td>957</td>\n",
       "      <td>6739</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:72</td>\n",
       "      <td>s1:i:889</td>\n",
       "      <td>dv:f:0.1062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>55</td>\n",
       "      <td>6046</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.29632</td>\n",
       "      <td>9799</td>\n",
       "      <td>302</td>\n",
       "      <td>6226</td>\n",
       "      <td>946</td>\n",
       "      <td>6066</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:69</td>\n",
       "      <td>s1:i:887</td>\n",
       "      <td>dv:f:0.1042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>85</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.17180</td>\n",
       "      <td>11608</td>\n",
       "      <td>3420</td>\n",
       "      <td>10764</td>\n",
       "      <td>938</td>\n",
       "      <td>7461</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:58</td>\n",
       "      <td>s1:i:878</td>\n",
       "      <td>dv:f:0.1188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>1218</td>\n",
       "      <td>7521</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR7825134.102167</td>\n",
       "      <td>12373</td>\n",
       "      <td>5408</td>\n",
       "      <td>11550</td>\n",
       "      <td>937</td>\n",
       "      <td>6353</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:58</td>\n",
       "      <td>s1:i:865</td>\n",
       "      <td>dv:f:0.1130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>882</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.130891</td>\n",
       "      <td>14700</td>\n",
       "      <td>6664</td>\n",
       "      <td>13189</td>\n",
       "      <td>937</td>\n",
       "      <td>6668</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:60</td>\n",
       "      <td>s1:i:865</td>\n",
       "      <td>dv:f:0.1131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>55</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.76456</td>\n",
       "      <td>12524</td>\n",
       "      <td>1553</td>\n",
       "      <td>8891</td>\n",
       "      <td>940</td>\n",
       "      <td>7497</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:62</td>\n",
       "      <td>s1:i:862</td>\n",
       "      <td>dv:f:0.1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>1417</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.108690</td>\n",
       "      <td>12058</td>\n",
       "      <td>87</td>\n",
       "      <td>6116</td>\n",
       "      <td>904</td>\n",
       "      <td>6114</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:52</td>\n",
       "      <td>s1:i:860</td>\n",
       "      <td>dv:f:0.1154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>85</td>\n",
       "      <td>7455</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR7825134.30530</td>\n",
       "      <td>12614</td>\n",
       "      <td>3930</td>\n",
       "      <td>11182</td>\n",
       "      <td>916</td>\n",
       "      <td>7419</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:59</td>\n",
       "      <td>s1:i:857</td>\n",
       "      <td>dv:f:0.1182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>55</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.30529</td>\n",
       "      <td>14545</td>\n",
       "      <td>3138</td>\n",
       "      <td>10545</td>\n",
       "      <td>903</td>\n",
       "      <td>7501</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:68</td>\n",
       "      <td>s1:i:848</td>\n",
       "      <td>dv:f:0.1129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>622</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.106363</td>\n",
       "      <td>13941</td>\n",
       "      <td>1245</td>\n",
       "      <td>8103</td>\n",
       "      <td>892</td>\n",
       "      <td>6945</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:59</td>\n",
       "      <td>s1:i:843</td>\n",
       "      <td>dv:f:0.1154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>85</td>\n",
       "      <td>7404</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.79502</td>\n",
       "      <td>23813</td>\n",
       "      <td>15497</td>\n",
       "      <td>22975</td>\n",
       "      <td>905</td>\n",
       "      <td>7521</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:54</td>\n",
       "      <td>s1:i:833</td>\n",
       "      <td>dv:f:0.1212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>729</td>\n",
       "      <td>7455</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR7825134.159776</td>\n",
       "      <td>13305</td>\n",
       "      <td>6592</td>\n",
       "      <td>13260</td>\n",
       "      <td>889</td>\n",
       "      <td>6800</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:51</td>\n",
       "      <td>s1:i:828</td>\n",
       "      <td>dv:f:0.1203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>85</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.64273</td>\n",
       "      <td>15867</td>\n",
       "      <td>4127</td>\n",
       "      <td>11510</td>\n",
       "      <td>858</td>\n",
       "      <td>7453</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:64</td>\n",
       "      <td>s1:i:817</td>\n",
       "      <td>dv:f:0.1151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>55</td>\n",
       "      <td>7404</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.142751</td>\n",
       "      <td>15244</td>\n",
       "      <td>2637</td>\n",
       "      <td>9951</td>\n",
       "      <td>865</td>\n",
       "      <td>7422</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:53</td>\n",
       "      <td>s1:i:815</td>\n",
       "      <td>dv:f:0.1219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>95</td>\n",
       "      <td>6474</td>\n",
       "      <td>-</td>\n",
       "      <td>SRR7825134.131064</td>\n",
       "      <td>22929</td>\n",
       "      <td>55</td>\n",
       "      <td>6394</td>\n",
       "      <td>867</td>\n",
       "      <td>6458</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:56</td>\n",
       "      <td>s1:i:812</td>\n",
       "      <td>dv:f:0.1148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR7825134.1</td>\n",
       "      <td>7563</td>\n",
       "      <td>55</td>\n",
       "      <td>7455</td>\n",
       "      <td>+</td>\n",
       "      <td>SRR7825134.20818</td>\n",
       "      <td>11224</td>\n",
       "      <td>2186</td>\n",
       "      <td>9710</td>\n",
       "      <td>877</td>\n",
       "      <td>7583</td>\n",
       "      <td>0</td>\n",
       "      <td>tp:A:S</td>\n",
       "      <td>cm:i:52</td>\n",
       "      <td>s1:i:808</td>\n",
       "      <td>dv:f:0.1229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0     1     2     3  4                  5      6      7      8   \\\n",
       "0   SRR7825134.1  7563   729  7455  +   SRR7825134.66305   9555   1321   8140   \n",
       "1   SRR7825134.1  7563    85  5480  +  SRR7825134.135869  13250   7604  12972   \n",
       "2   SRR7825134.1  7563   729  7404  +   SRR7825134.50128  16210    653   7265   \n",
       "3   SRR7825134.1  7563    97  7455  -    SRR7825134.7296  13396   5948  13367   \n",
       "4   SRR7825134.1  7563    55  6643  -  SRR7825134.119454  10268    218   6869   \n",
       "5   SRR7825134.1  7563    55  6046  +   SRR7825134.29632   9799    302   6226   \n",
       "6   SRR7825134.1  7563    85  7455  +   SRR7825134.17180  11608   3420  10764   \n",
       "7   SRR7825134.1  7563  1218  7521  -  SRR7825134.102167  12373   5408  11550   \n",
       "8   SRR7825134.1  7563   882  7455  +  SRR7825134.130891  14700   6664  13189   \n",
       "9   SRR7825134.1  7563    55  7455  +   SRR7825134.76456  12524   1553   8891   \n",
       "10  SRR7825134.1  7563  1417  7455  +  SRR7825134.108690  12058     87   6116   \n",
       "11  SRR7825134.1  7563    85  7455  -   SRR7825134.30530  12614   3930  11182   \n",
       "12  SRR7825134.1  7563    55  7455  +   SRR7825134.30529  14545   3138  10545   \n",
       "13  SRR7825134.1  7563   622  7455  +  SRR7825134.106363  13941   1245   8103   \n",
       "14  SRR7825134.1  7563    85  7404  +   SRR7825134.79502  23813  15497  22975   \n",
       "15  SRR7825134.1  7563   729  7455  -  SRR7825134.159776  13305   6592  13260   \n",
       "16  SRR7825134.1  7563    85  7455  +   SRR7825134.64273  15867   4127  11510   \n",
       "17  SRR7825134.1  7563    55  7404  +  SRR7825134.142751  15244   2637   9951   \n",
       "18  SRR7825134.1  7563    95  6474  -  SRR7825134.131064  22929     55   6394   \n",
       "19  SRR7825134.1  7563    55  7455  +   SRR7825134.20818  11224   2186   9710   \n",
       "\n",
       "      9     10  11      12       13         14           15  \n",
       "0   1083  6903   0  tp:A:S  cm:i:74  s1:i:1008  dv:f:0.1064  \n",
       "1    995  5486   0  tp:A:S  cm:i:69   s1:i:935  dv:f:0.1004  \n",
       "2    970  6743   0  tp:A:S  cm:i:54   s1:i:913  dv:f:0.1179  \n",
       "3    950  7458   0  tp:A:S  cm:i:59   s1:i:910  dv:f:0.1181  \n",
       "4    957  6739   0  tp:A:S  cm:i:72   s1:i:889  dv:f:0.1062  \n",
       "5    946  6066   0  tp:A:S  cm:i:69   s1:i:887  dv:f:0.1042  \n",
       "6    938  7461   0  tp:A:S  cm:i:58   s1:i:878  dv:f:0.1188  \n",
       "7    937  6353   0  tp:A:S  cm:i:58   s1:i:865  dv:f:0.1130  \n",
       "8    937  6668   0  tp:A:S  cm:i:60   s1:i:865  dv:f:0.1131  \n",
       "9    940  7497   0  tp:A:S  cm:i:62   s1:i:862  dv:f:0.1163  \n",
       "10   904  6114   0  tp:A:S  cm:i:52   s1:i:860  dv:f:0.1154  \n",
       "11   916  7419   0  tp:A:S  cm:i:59   s1:i:857  dv:f:0.1182  \n",
       "12   903  7501   0  tp:A:S  cm:i:68   s1:i:848  dv:f:0.1129  \n",
       "13   892  6945   0  tp:A:S  cm:i:59   s1:i:843  dv:f:0.1154  \n",
       "14   905  7521   0  tp:A:S  cm:i:54   s1:i:833  dv:f:0.1212  \n",
       "15   889  6800   0  tp:A:S  cm:i:51   s1:i:828  dv:f:0.1203  \n",
       "16   858  7453   0  tp:A:S  cm:i:64   s1:i:817  dv:f:0.1151  \n",
       "17   865  7422   0  tp:A:S  cm:i:53   s1:i:815  dv:f:0.1219  \n",
       "18   867  6458   0  tp:A:S  cm:i:56   s1:i:812  dv:f:0.1148  \n",
       "19   877  7583   0  tp:A:S  cm:i:52   s1:i:808  dv:f:0.1229  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://raw.githubusercontent.com/genomics-course/11-assembly/master/notebooks/SRR7825134.20lines.paf\"\n",
    "pd.read_csv(url, sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Layout: assemble contigs from overlapping reads\n",
    "The program `miniasm` is an OLC assembler, it takes the raw long-read data (.fastq) and the mapping information (.paf) and constructs contigs by constructing a graph that connects sequences based on the overlaps. An output is produced in [graphical fragment assembly format (.gfa)](https://github.com/GFA-spec/GFA-spec). This format can be used in some software (e.g. [Bandage](http://rrwick.github.io/Bandage)) to visualize the assembly. It contains within it the assembled contig sequence information as well as the loops and bubbles (ambiguities) in their assembly (the graph)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# miniasm -f SRR7825134.fastq SRR7825134.paf > SRR7825134.gfa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# visualize the assembly graph using Bandage\n",
    "wget https://github.com/rrwick/Bandage/releases/download/v0.8.1/Bandage_Ubuntu_dynamic_v0_8_1.zip\n",
    "unzip Bandage_Ubuntu_dynamic_v0_8_1.zip\n",
    "\n",
    "Bandage image SRR7825134.gfa SRR7825134.gfa.png --names --lengths --outline 0 --colour random --fontsize 10 --iter 4\n",
    "\n",
    "from IPython.display import Image\n",
    "Image(filename='SRR7825134.gfa.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we just want the best estimate of the assembly we can extract the contig sequences from this file to get a fasta file. This can be done using the unix tool `awk` to extract the lines from the `.gfa` file that start with \"S\" and convert them to fasta format, like bellow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# awk '$1 ~/S/ {print \">\"$2\"\\n\"$3}' SRR7825134.gfa > assembly_miniasm.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Consensus: polish the assembled contigs by mapping reads again\n",
    "\n",
    "Technically the last step by `miniasm` performs both a *layout* and *consensus* step of the OLC assembly. The assembly fasta file was constructed from the overlaps among the long reads. However, it is common with long-read assemblies to apply one or more additional *consensus* steps that aim to polish the final base calls. Example programs for final polishing include `racon` and `quiver` which use PacBio reads to polish the data, or `pilon` which can use Illumina data to polish it (The lastest version of `pilon` added the support for using long reads). We will skip this step for now since it is often very time consuming. \n",
    "\n",
    "Before we analyze this genome file, let's try assembling the genome again from PacBio data using a different program. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Canu assembly\n",
    "The program [Canu](https://canu.readthedocs.io/en/latest/quick-start.html) is considered one of the best long-read assemblers, but it is much slower than the last approach we tried. While the assembly above using miniasm finished in just a few minutes, the canu assembly below took several hours. Once we get to the assembly comparisons you can decide if you think the longer run time was worth it.\n",
    "\n",
    "One of the reasons canu is much slower is that it performs the full suite of steps we've discussed: pre-cleanup of reads, read overlap calculations, layout graph construction, consensus contig construction, and post-analysis contig polishing. A nice feature of canu is that you can tell it to do all of this with a single command rather than requiring multiple different programs or commands. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# canu parameters explained \n",
    "# -p: prefix name for output files\n",
    "# -d: output directory name\n",
    "# -genomeSize: estimate of size\n",
    "# -pacbio-raw: the raw fastq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# canu -p SRR7825124 -d assembly_canu genomeSize=30m -pacbio-raw SRR7825134.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While canu is running it prints a long log file to the screen explaining each step of the analysis and how much resources it is consuming, very similar to spades in the last notebook. We'll skip looking at the log file for now though and continue to comparing the resulting files. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare all four genome assemblies\n",
    "Again we will use quast to assess the quality of our genome assemblies. The command below inputs the assembled contigs in fasta format of all four assemblies we've completed using four different approaches. It measures summary statistics of assembly quality and also BUSCO scores. You can read the report [here](https://genomics-course.github.io/11-assembly/notebooks/comparison/report.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# quast.py \\\n",
    "#     assembly_spades/scaffolds.fasta \\\n",
    "#     assembly_spades_hybrid/scaffolds.fasta \\\n",
    "#     assembly_miniasm.fasta \\\n",
    "#     assembly_canu/SRR7825124.fasta \\\n",
    "#     -o comparison \\\n",
    "#     --conserved-genes-finding \\ \n",
    "#     --fungus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-success\">\n",
    "    <b>[8] Question:</b> \n",
    "    Which of the four assembly methods yielded the best assembly in your opinion? Which statistics did you base this off of? Why do you think the contig N50 was higher in some assemblies versus others? Why do you think the complete BUSCO score was higher in some versus others? Answer in Markdown below.   \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems overall that the `canu` long-read assembly is the best. This is based on the number of contigs (100), N50 (860Kb), and BUSCO scores (90%). The longer reads in pacbio assemblies allow them to connect many smaller fragments together to get a larger N50. The canu assembly probably does best because it polishes the long-reads before finding their overlaps. I think that the high error rate in the pacbio only assemblies might make it hard to identify homology with the single-copy orthologs in BUSCO, and so they have lower scores. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-success\">\n",
    "    <b>[9] Question:</b> \n",
    "    What assembly software did Sumpter et al. use in the paper to assemble the *Bd* genome? How do their assembly statistics compare to ours? Why do you think they might differ?\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sumpter et al. used Canu. Our canu assembly actually has a better N50 and fewer contigs than theirs. This might be because the canu software has improved and we have a newer version. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You just completed several genome assemblies that in total took only a few hours. Are you surprised by how easy or difficult this was? The time and expense involved with attaining the data is the hardest part, followed by getting sufficient computational resources to assemble the genome. But overall, it's not a super difficult task. The `canu` and `spades` assemblies could be done with a single command to the programs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-success\">\n",
    "    <b>[10] Question:</b> \n",
    "    Search on google scholar for another published genome from within the last three years. In the Markdown cell below report the citation for the publication, how large the genome is, which technology was used to sequence it, and a summary of the genome assembly (e.g., N50 and/or number of scaffolds/contigs). Try to find an example for an organism that is of interest to you.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
